From 7ecd5d14dc11e79974b3cfdad08262f154be7c0d Mon Sep 17 00:00:00 2001
From: Robin Gong <yibin.gong@nxp.com>
Date: Thu, 5 Jan 2017 14:15:17 +0800
Subject: [PATCH 1383/5242] MLK-13720-4: regulator: pf1550-regulator-rpmsg:
 fix "rpmsg_send timeout"

commit  f52eb75149fe3c0a1ba0a0aed1db244788e5d911 from
https://source.codeaurora.org/external/imx/linux-imx.git

Sometimes rpmsg callback triggered quickly before reinit_completion, then
cause "rpmsg_send timeout!". Move reinit_completion to the place before
rpmsg_send to make sure the completion is ready before callback triggered.

Signed-off-by: Robin Gong <yibin.gong@nxp.com>
(cherry picked from commit 8e44277d3f27e5f46232bd705ee9ef594db42575)
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/regulator/pf1550-regulator-rpmsg.c |    5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/drivers/regulator/pf1550-regulator-rpmsg.c b/drivers/regulator/pf1550-regulator-rpmsg.c
index fd0f69d..13f4484 100644
--- a/drivers/regulator/pf1550-regulator-rpmsg.c
+++ b/drivers/regulator/pf1550-regulator-rpmsg.c
@@ -101,6 +101,9 @@ static int pf1550_send_message(struct pf1550_regulator_rpmsg *msg,
 
 	pm_qos_add_request(&info->pm_qos_req, PM_QOS_CPU_DMA_LATENCY, 0);
 
+	/* wait response from rpmsg */
+	reinit_completion(&info->cmd_complete);
+
 	err = rpmsg_send(info->rpdev->ept, (void *)msg,
 			    sizeof(struct pf1550_regulator_rpmsg));
 
@@ -110,8 +113,6 @@ static int pf1550_send_message(struct pf1550_regulator_rpmsg *msg,
 		dev_err(&info->rpdev->dev, "rpmsg_send failed: %d\n", err);
 		return err;
 	}
-	/* wait response from rpmsg */
-	reinit_completion(&info->cmd_complete);
 	err = wait_for_completion_timeout(&info->cmd_complete,
 					  msecs_to_jiffies(RPMSG_TIMEOUT));
 	if (!err) {
-- 
1.7.9.5

